Fixes #26805 - react-router foundation #4

glekner opened the pull request:
- add server props to every route ( are there any more props a route should have ? )
- write a helper to get searchBar Props. @Laviro please take a look
- add pagination props (disablePrev, disableNext)

<!--
Simple description of what is fixed/introduced.

Prerequisites for testing:
* VMware cluster
* Two smart proxies

Tests needed (anticipated impacts):
* Hosts list - mainly power column
* Power status on host page
-->

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
* Suggest prerequisites for testing and testing scenarios following example above.
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

-->

sharvit commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -1,23 +1,22 @@
           currentLocation = renderProps.location;
 
-          return render(renderProps);
+          return render({ ...renderProps, ...serverProps });
```
Does it mean only routes will receive the `metadata` props?

imo it should be an `application-metadata` reducer instead so all components get the app-metadata info by using selectors.

sharvit commented on webpack/assets/javascripts/react_app/components/Pagination/Pagination.js:
```diff
@@ -17,6 +17,8 @@ const Pagination = props => {
     onPageSet,
     onPerPageSelect,
     dropdownButtonId,
+    disableNext,
```
Is this apart of the scope? Can it be done in a different PR?

glekner replied:
its part of the scope yes, adding foundation for router pagination.

Ron-Lavi commented on webpack/assets/javascripts/react_app/constants.js:
```diff
@@ -3,3 +3,22 @@ export const STATUS = {
+  controller,
+  autocomplete: {
+    id,
+    searchQuery: '',
```
`searchQuery` should be able to update from server

glekner replied:
`searchQuery` no longer gets updates from the server :) thats the point
with new views implementing the router approach, there won't be full page reloads.
If theres a full page reload, the page will fetch it from the URI on mount.

Ron-Lavi commented on webpack/assets/javascripts/react_app/constants.js:
```diff
@@ -3,3 +3,22 @@ export const STATUS = {
+  autocomplete: {
+    id,
+    searchQuery: '',
+    url: `${controller}/auto_complete_search`,
```
I think `url` should be also passed from server,
unless you would like to mimic: `send("auto_complete_search_#{auto_complete_controller_name}_path")`

glekner replied:
same as the comment above, there are no longer props from the server for routes. do we want to avoid an API call here? @ohadlevy

Ron-Lavi commented on webpack/assets/javascripts/react_app/constants.js:
```diff
@@ -3,3 +3,22 @@ export const STATUS = {
+  bookmarks: {
+    url: '/api/bookmarks',
+    canCreate,
+    documentationUrl: `4.1.5Searching`,
```
won't you prefer to get the documentation `url` ready from server?
`documentation_url("4.1.5Searching")`

glekner replied:
again, no props from the server. thats why its now in redux

Ron-Lavi replied:
in the last comments I meant using / implementing those methods that are used in the server.
Receiving them as `metadata` is also optional, isn't it ?

glekner replied:
instead of maintaining the functions in two places we receive it [here](https://github.com/theforeman/foreman/pull/6767/files#diff-127a850b7d7b20173ad3fba88f8a25e3R532) for now.
could be good to move the implementation to the client later

Ron-Lavi commented on webpack/assets/javascripts/react_app/constants.js:
```diff
@@ -3,3 +3,22 @@ export const STATUS = {
+  },
+  bookmarks: {
+    url: '/api/bookmarks',
+    canCreate,
```
maybe rename to `canCreateBookmarks` ?

glekner commented:
_summarizing changes_
- add server props to every route ( are there any more props a route should have ? )
- write a helper to get searchBar Props. @Laviro please take a look
- add pagination props (disablePrev, disableNext)
- added a Reducer and Selectors for ReactApp, now contains a metadata state. The Layout adds server props to the state onMount
- added `withRenderHandler` HOC - Helps render the right component based on state (isLoading, hasError, hasData)
- added common EmptyPage, LoadingPage
- added `recompose` to package.json for future use (Lifecycle HOCs)

glekner commented:
@sharvit - I prefer to leave the `withRenderHandler` like this, giving it a `onMount` arg can be confusing because we wrap it with a functional component thats inside a `PageLayout` wrapper. importing an action from that inner component and passing it to the HOC seems out of place to me.

ohadlevy commented on package.json:
```diff
@@ -108,6 +108,7 @@
     "react-password-strength": "^2.4.0",
     "react-redux": "^5.1.1",
     "react-router-dom": "^4.3.1",
+    "recompose": "^0.30.0",
```
packaging PR please?

glekner replied:
https://github.com/theforeman/foreman-packaging/pull/3825

ohadlevy replied:
btw where is this used?

glekner replied:
not used yet here, only in the blocked PRs

ohadlevy replied:
hmm.. taken from https://github.com/acdlite/recompose

```
A Note from the Author (acdlite, Oct 25 2018):
Hi! I created Recompose about three years ago. About a year after that, I joined the React team. Today, we announced a proposal for Hooks. Hooks solves all the problems I attempted to address with Recompose three years ago, and more on top of that. I will be discontinuing active maintenance of this package (excluding perhaps bugfixes or patches for compatibility with future React releases), and recommending that people use Hooks instead. Your existing code with Recompose will still work, just don't expect any new features. Thank you so, so much to @wuct and @istarkov for their heroic work maintaining Recompose over the last few years.

Read more discussion about this decision here.
```

glekner committed:
Fixes #26805 - react-router foundation #4

ohadlevy commented:
LGTM - @sharvit any additional feedback?

glekner commented:
- removed `recompose`
- added `HOC.js` inside routes folder, added `withRenderHandler` and `callOnMount` to it
- `callOnMount` now uses `useEffect` instead of `recompose`

glekner commented:
added `callOnPopState` to `HOC.js`
assuming the component uses `withRouter`, you can use `callOnPopState` to trigger a function on browser navigation (popstate)

glekner commented:
@sharvit can you review?

Ron-Lavi reviewed (commented):
Looks great apart from the `getControllerSearchProps` as consts.

sharvit commented on webpack/assets/javascripts/react_app/ReactApp/ReactAppHelper.js:
```diff
@@ -0,0 +1,5 @@
+import store from '../redux';
+import { APP_FETCH_SERVER_PROPS } from '../constants';
+
+export const initializeMetadata = metadata =>
```
Can't it be an action? It feels like a workaround redux when it can simply be implemented with redux best practices.

sharvit commented on webpack/assets/javascripts/react_app/constants.js:
```diff
@@ -1,5 +1,26 @@
+export const APP_FETCH_SERVER_PROPS = 'APP_FETCH_SERVER_PROPS';
```
IMO it should belong to `ReactAppConstants.js`

sharvit commented on webpack/assets/javascripts/react_app/routes/HOC.js:
```diff
@@ -0,0 +1,61 @@
+import React, { useEffect } from 'react';
```
This file contains really nice HOCs :+1: 

I'm wondering how the documentation part of it should be handled, especially when there are no use cases yet.
I believe adding usage examples in the comments would do the job.
Another option is to update the storybook with some examples.
Or move it to another PR that use this infrastructure.

sharvit replied:
Can you write tests for the HOCs

sharvit reviewed (changes requested):
Thanks @glekner, I like the direction you are taking us :+1:

glekner commented:
@sharvit made ReactApp connected with actions.

About the HOCs, I already know of potential use-cases for all of them :) we should work on good documentation and deep dives.

sharvit commented on webpack/assets/javascripts/react_app/ReactApp/ReactApp.js:
```diff
@@ -0,0 +1,34 @@
+  initializeMetadata,
+  stateMetadata,
+}) => {
+  if (isEmpty(stateMetadata)) initializeMetadata(metadata);
```
Shouldn't it happen only `onMount`?

glekner replied:
this is to prevent the action if not needed?

sharvit commented on webpack/assets/javascripts/react_app/ReactApp/ReactAppActions.js:
```diff
@@ -0,0 +1,4 @@
+import { APP_FETCH_SERVER_PROPS } from './ReactAppConstants';
+
+export const initializeMetadata = metadata => dispatch =>
+  dispatch({ type: APP_FETCH_SERVER_PROPS, payload: metadata });
```
`dispatch` is redundant.

sharvit commented:
> About the HOCs, I already know of potential use-cases for all of them :) we should work on good documentation and deep dives.

+1

Can you just add simple usage example in the comments?

glekner commented:
- removed `onPopState` HOC for now
- added tests for the HOCs
- using `callOnMount` on `ReactApp` 🚀

sharvit reviewed (approved):
LGTM 👍 Thanks @glekner 

It seems that failing tests are not relevant, but I'm not 100% sure.
